### PR TITLE
Great filters improvement

### PR DESCRIPTION
**Pull request by FerarDuanSednan**
- Allow filter arguments
- Add filters: upcase, minus, replace
- Filter can now work with Value type and not only raw str ( allow filter on arrays, objects, etc )
- changes 100% tested

### CODE REVIEW

File: src/filters.rs [hunk 1]
```diff
@@ -1,9 +1,121 @@
...
+    }
+}
+
+pub type FilterResult = Result<String, FilterError>;
+
+pub fn size(input : &Value, _args: &Vec<Value>) -> FilterResult {
+    match input {
+        &Str(ref s)    => Ok(s.len().to_string()),
+        &Array(ref x)  => Ok(x.len().to_string()),
+        &Object(ref x) => Ok(x.len().to_string()),
```
Comment (johannhof) [inline src/filters.rs]: you can abbreviate this to `&Str(ref s) | &Array(ref x) | &Object(ref x) => Ok(s.len().to_string())` if I'm not mistaken

File: README.md [hunk 2]
```diff
@@ -17,3 +17,59 @@ Now you can use the crate in your code
 ```rust
 extern crate liquid;
 ```
+Exemple:
```
Comment (johannhof) [inline README.md]: Exemple => Example :)

File: README.md [hunk 3]
```diff
@@ -17,3 +17,59 @@ Now you can use the crate in your code
...
+let mut data = Context::new();
+data.set_val("num", Value::Num(5f32));
+data.set_val("numTwo", Value::Num(6f32));
+
+let output = template.render(&mut data);
+```
+
+Plugins
+--------
+Cache block ( File and Redis ) : https://github.com/FerarDuanSednan/liquid-rust-cache
```
Comment (johannhof) [inline README.md]: :+1:

File: src/filters.rs [hunk 4]
```diff
@@ -1,9 +1,121 @@
...
+            let arg2 = match &args[1] {
+                &Str(ref a) => a, _ => return Err(InvalidArgument(1, "Str expected".to_string()))
+            };
+            Ok(x.replace(arg1, arg2))
+        },
+        _ => Err(InvalidType("String expected".to_string()))
+    }
+}
+
+macro_rules! unit {
```
Comment (johannhof) [inline src/filters.rs]: The use of macros is clever, but please wrap everything (line 85-end) in a [tests module](https://doc.rust-lang.org/book/testing.html#the-tests-module). This makes sure it's clear semantically that it's not used anywhere else and helps the compiler.

File: src/template.rs [hunk 5]
```diff
@@ -10,6 +10,9 @@ pub struct Template<'a> {
 impl<'a> Renderable for Template<'a> {
     fn render(&self, context: &mut Context) -> Result<Option<String>> {
         context.filters.insert("size".to_string(), Box::new(size));
+        context.filters.insert("upcase".to_string(), Box::new(upcase));
+        context.filters.insert("minus".to_string(), Box::new(minus));
+        context.filters.insert("replace".to_string(), Box::new(replace));
```
Comment (johannhof) [inline src/template.rs]: We  really need to find a better way to do this than create everything at render eventually, but that's not for this PR :sweat_smile:

**Comment by johannhof** [comment]
This is a great pull request, thanks! I already assumed that transforming strings might no be sufficient for filters. Thanks for changing that. Now implementing more filters will be much easier.

Please correct the minor things I found and we should be ready to go :rocket:

File: src/filters.rs (hunk 1 above)
Comment (FerarDuanSednan) [inline src/filters.rs]: I tried like this:
```rust
&Str(ref x) |&Array(ref x) | &Object(ref x) => Ok(x.len().to_string())
```
But it fail with: _error: mismatched types:_

**Comment by FerarDuanSednan** [comment]
I made the changes.

I think a part of the work I did (commit 53a9655) is related to #3 ( for the use with the iron framework : [ iron-liquid ](https://github.com/FerarDuanSednan/iron-liquid) ).

**Comment by johannhof** [comment]
I suspected as much, very nice! One more thing: I'd love if you could squash this into fewer commits. It doesn't have to be one commit, but please at least get rid of duplicate `Update README.md` and the merge commit. If you need help with rebasing I'd be happy to help you!

**Comment by FerarDuanSednan** [comment]
I made some mistakes during the process but it seems everything is ok now :p

**Comment by johannhof** [comment]
Looks almost perfect! Sorry, there's on more thing I'd like to improve. :snowman:

The first three commits are duplicates of existing commits, but with different hashes (probably happened while rebasing or whatever). Please get rid of them, it would spam the commit history. That shouldn't be too hard I hope, but if you need assistance I'm here :)

**Comment by FerarDuanSednan** [comment]
Done :sparkles:

File: src/parser.rs [hunk 6]
```diff
@@ -76,9 +76,14 @@ fn parse_output<'a>(tokens: &Vec<Token>) -> Result<Box<Renderable + 'a>> {
             return Err(Error::Parser("parse_output: expected a colon".to_owned()));
         }
 
+        iter.next(); // skip colon
+
+        println!("{:?}", tokens);
```
Comment (johannhof) [inline src/parser.rs]: Didn't notice that debug print statement before, this is probably not supposed to be there, right? :)

**Comment by johannhof** [comment]
Perfect, thanks! You have a `println!` left in your code I fear :fearful:, but apart from that it LGTM!

**Comment by FerarDuanSednan** [comment]
ups  :scream:, removing this right now  :innocent:

**Comment by johannhof** [comment]
:+1: Thanks for hanging in, great contribution!

**Pull request merged**